File: backend/src/aidetect/schemas/detection.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field

Mode = Literal["fast", "balanced", "thorough"]
Verdict = Literal["human", "likely_human", "uncertain", "likely_ai", "ai"]
Severity = Literal["low", "medium", "high"]
ChunkStrategy = Literal["none", "sentence", "paragraph", "auto"]


class DetectionOptions(BaseModel):
    include_evidence: bool = True
    include_signals: bool = True
    include_rubric_scores: bool = False
    chunk_strategy: ChunkStrategy = "auto"


class DetectionContext(BaseModel):
    platform: str | None = None
    expected_register: str | None = None


class DetectionRequest(BaseModel):
    model_config = ConfigDict(str_strip_whitespace=False)

    text: str = Field(..., min_length=1)
    language: str = "auto"
    mode: Mode = "balanced"
    options: DetectionOptions = Field(default_factory=DetectionOptions)
    context: DetectionContext | None = None
    rubric_version: str = "latest"
    model_version: str = "latest"


class PatternMatch(BaseModel):
    category: str
    name: str
    span: tuple[int, int]
    severity: Severity = "medium"
    suggestion: str | None = None


class StatisticalSignal(BaseModel):
    score: float
    contributors: dict[str, object] = Field(default_factory=dict)


class PatternsSignal(BaseModel):
    score: float
    matches: list[PatternMatch] = Field(default_factory=list)


class SentenceScore(BaseModel):
    index: int
    text: str
    score: float
    reason: str | None = None


class LLMJudgeSignal(BaseModel):
    score: float
    model: str
    summary: str | None = None
    rubric_scores: dict[str, "RubricDimensionScore"] | None = None
    error: str | None = None
    sentence_scores: list[SentenceScore] | None = None
    sentence_aggregate: dict[str, float | int] | None = None
    usage: dict[str, int] | None = None


class Signals(BaseModel):
    statistical: StatisticalSignal | None = None
    patterns: PatternsSignal | None = None
    llm_judge: LLMJudgeSignal | None = None


class RubricDimensionScore(BaseModel):
    score: float
    evidence: str | None = None


class EvidenceSpan(BaseModel):
    start: int
    end: int
    type: str
    severity: Severity = "medium"
    reason: str


class Evidence(BaseModel):
    spans: list[EvidenceSpan] = Field(default_factory=list)
    annotated_html: str | None = None
    annotated_markdown: str | None = None


class DetectionResultBlock(BaseModel):
    ai_probability: float
    human_probability: float
    verdict: Verdict
    confidence: float
    severity: Severity


class RequestDetails(BaseModel):
    text_hash: str
    length_chars: int
    length_tokens: int | None = None
    language_detected: str | None = None


class ResponseMetadata(BaseModel):
    model_version: str
    rubric_version: str
    mode: Mode
    duration_ms: int
    cached: bool = False
    cost_credits: int = 1


class DetectionResponse(BaseModel):
    id: UUID
    object: Literal["detection"] = "detection"
    created_at: datetime
    result: DetectionResultBlock
    signals: Signals | None = None
    rubric_scores: dict[str, RubricDimensionScore] | None = None
    evidence: Evidence | None = None
    request: RequestDetails
    metadata: ResponseMetadata

File: frontend/src/types/detection.ts
// Mirrors backend Pydantic schemas in src/aidetect/schemas/detection.py.
// Keep in sync when the API contract evolves.

export type Verdict = 'human' | 'likely_human' | 'uncertain' | 'likely_ai' | 'ai';
export type Severity = 'low' | 'medium' | 'high';
export type Mode = 'fast' | 'balanced' | 'thorough';
export type ChunkStrategy = 'none' | 'sentence' | 'paragraph' | 'auto';

export interface PatternMatch {
  category: string;
  name: string;
  span: [number, number];
  severity: Severity;
  suggestion: string | null;
}

export interface BurstinessContrib {
  sd: number | null;
  sentence_count: number;
  human_score?: number;
  flag?: string;
}

export interface TtrContrib {
  ttr: number | null;
  unique_words?: number;
  total_words?: number;
  in_range?: boolean;
  human_score?: number;
  flag?: string;
}

export interface SentenceCvContrib {
  cv: number | null;
  sd?: number;
  mean?: number;
  human_score?: number;
  flag?: string;
}

export interface FormattingContrib {
  bold_md_count: number;
  emoji_count: number;
  emdash_count: number;
  has_unicode_bold: boolean;
  hashtag_count: number;
  violations: number;
  flags: string[];
  human_score: number;
}

export interface StatisticalSignal {
  score: number;
  contributors: {
    human_score?: number;
    burstiness?: BurstinessContrib;
    ttr?: TtrContrib;
    sentence_cv?: SentenceCvContrib;
    formatting?: FormattingContrib;
    flag?: string;
  };
}

export interface PatternsSignal {
  score: number;
  matches: PatternMatch[];
}

export interface SentenceScore {
  index: number;
  text: string;
  score: number;
  reason: string | null;
}

export interface SentenceAggregate {
  count: number;
  mean: number;
  max: number;
  fraction_ai_like: number;
}

export interface LLMUsage {
  prompt_tokens?: number;
  completion_tokens?: number;
  total_tokens?: number;
}

export interface LLMJudgeSignal {
  score: number;
  model: string;
  summary: string | null;
  error?: string | null;
  sentence_scores?: SentenceScore[] | null;
  sentence_aggregate?: SentenceAggregate | null;
  usage?: LLMUsage | null;
}

export interface Signals {
  statistical?: StatisticalSignal;
  patterns?: PatternsSignal;
  llm_judge?: LLMJudgeSignal;
}

export interface RubricDimensionScore {
  score: number;
  evidence: string | null;
}

export interface EvidenceSpan {
  start: number;
  end: number;
  type: string;
  severity: Severity;
  reason: string;
}

export interface Evidence {
  spans: EvidenceSpan[];
  annotated_html?: string | null;
  annotated_markdown?: string | null;
}

export interface DetectionResultBlock {
  ai_probability: number;
  human_probability: number;
  verdict: Verdict;
  confidence: number;
  severity: Severity;
}

export interface RequestDetails {
  text_hash: string;
  length_chars: number;
  length_tokens?: number | null;
  language_detected?: string | null;
}

export interface ResponseMetadata {
  model_version: string;
  rubric_version: string;
  mode: Mode;
  duration_ms: number;
  cached: boolean;
  cost_credits: number;
}

export interface DetectionResult {
  id: string;
  object: 'detection';
  created_at: string;
  result: DetectionResultBlock;
  signals?: Signals;
  rubric_scores?: Record<string, RubricDimensionScore>;
  evidence?: Evidence;
  request: RequestDetails;
  metadata: ResponseMetadata;
}

export interface DetectionOptions {
  include_evidence?: boolean;
  include_signals?: boolean;
  include_rubric_scores?: boolean;
  chunk_strategy?: ChunkStrategy;
}

export interface DetectionContext {
  platform?: string;
  expected_register?: string;
}

export interface DetectionRequest {
  text: string;
  language?: string;
  mode?: Mode;
  options?: DetectionOptions;
  context?: DetectionContext;
  rubric_version?: string;
  model_version?: string;
}
